File: examples/multi-period.py
"""
Balance Engine - Multi-Period Production Planning Example

This script demonstrates how to solve a multi-period production planning problem
using linear programming optimization.
"""

import pulp
import matplotlib.pyplot as plt
import numpy as np
import engine


def main():
    """Main demonstration function"""
    engine.init()  # Initialize the engine module

    print("Balance Engine - Multi-Period Production Planning Demo")

    # Define the problem data
    products = ["A", "B"]
    periods = ["January", "February", "March"]

    # Initial conditions
    initial_inventory = {"A": 100, "B": 120}
    safety_stock = {"A": 130, "B": 110}
    production_cost = {"A": 20, "B": 25}  # in euros
    holding_cost_rate = 0.02  # 2% of production cost

    # Calculate holding cost per unit per period
    holding_cost = {p: production_cost[p] * holding_cost_rate for p in products}

    # Demand forecasts
    demand = {
        ("A", "January"): 700,
        ("A", "February"): 900,
        ("A", "March"): 1000,
        ("B", "January"): 800,
        ("B", "February"): 600,
        ("B", "March"): 900,
    }

    # Resource capacities
    machine_capacity = {"January": 3000, "February": 2800, "March": 3600}
    labor_capacity = {"January": 2500, "February": 2300, "March": 2400}

    # Resource requirements per unit
    machine_hours = {"A": 1.5, "B": 1.6}
    labor_hours = {"A": 1.1, "B": 1.2}

    # Create the optimization model
    model = pulp.LpProblem(
        name="Multi_Period_Production_Planning", sense=pulp.LpMinimize
    )

    # Decision variables
    # x[p,t] = production of product p in period t
    x = {
        (p, t): pulp.LpVariable(f"x_{p}_{t}", lowBound=0, cat=pulp.LpInteger)
        for p in products
        for t in periods
    }

    # inv[p,t] = inventory of product p at the end of period t
    inv = {
        (p, t): pulp.LpVariable(f"inv_{p}_{t}", lowBound=0, cat=pulp.LpInteger)
        for p in products
        for t in periods
    }

    # Objective function: minimize total cost (production + inventory holding)
    total_production_cost = pulp.lpSum(
        [production_cost[p] * x[p, t] for p in products for t in periods]
    )
    total_holding_cost = pulp.lpSum(
        [holding_cost[p] * inv[p, t] for p in products for t in periods]
    )

    model += total_production_cost + total_holding_cost, "Total Cost"

    # Constraints
    # 1. Inventory balance constraints
    for p in products:
        for i, t in enumerate(periods):
            if i == 0:  # first period
                model += (
                    inv[p, t] == initial_inventory[p] + x[p, t] - demand[p, t],
                    f"Inventory_Balance_{p}_{t}",
                )
            else:
                prev_t = periods[i - 1]
                model += (
                    inv[p, t] == inv[p, prev_t] + x[p, t] - demand[p, t],
                    f"Inventory_Balance_{p}_{t}",
                )

    # 2. Capacity constraints
    for t in periods:
        # Machine capacity
        model += (
            pulp.lpSum([machine_hours[p] * x[p, t] for p in products])
            <= machine_capacity[t],
            f"Machine_Capacity_{t}",
        )
        # Labor capacity
        model += (
            pulp.lpSum([labor_hours[p] * x[p, t] for p in products])
            <= labor_capacity[t],
            f"Labor_Capacity_{t}",
        )

    # 3. Safety stock requirements (end of planning horizon)
    for p in products:
        model += inv[p, periods[-1]] >= safety_stock[p], f"Safety_Stock_{p}"

    # Solve the model
    model.solve()

    print(f"Status: {pulp.LpStatus[model.status]}")

    if model.status == pulp.LpStatusOptimal:
        print("\nOptimal Solution Found:")
        print(f"Total Cost: €{pulp.value(model.objective):.2f}")

        # Print production plan
        print("\nProduction Plan:")
        print(f"{'Period':<10} {'Product A':<10} {'Product B':<10}")
        print("-" * 30)
        for t in periods:
            print(f"{t:<10} {int(x['A', t].value()):<10} {int(x['B', t].value()):<10}")

        # Print ending inventory
        print("\nEnding Inventory:")
        print(f"{'Period':<10} {'Product A':<10} {'Product B':<10}")
        print("-" * 30)
        for t in periods:
            print(
                f"{t:<10} {int(inv['A', t].value()):<10} {int(inv['B', t].value()):<10}"
            )

        # Calculate resource utilization
        print("\nResource Utilization:")
        print(
            f"{'Period':<10} {'Machine Hours (Used/Available)':<30} {'Labor Hours (Used/Available)':<30}"
        )
        print("-" * 70)
        for t in periods:
            machine_used = sum(machine_hours[p] * x[p, t].value() for p in products)
            machine_util = machine_used / machine_capacity[t] * 100
            labor_used = sum(labor_hours[p] * x[p, t].value() for p in products)
            labor_util = labor_used / labor_capacity[t] * 100
            print(
                f"{t:<10} {machine_used:.1f}/{machine_capacity[t]} ({machine_util:.1f}%) {' ' * 5} {labor_used:.1f}/{labor_capacity[t]} ({labor_util:.1f}%)"
            )

        # Visualize the results
        plot_results(periods, products, x, inv, demand)
    else:
        print("No optimal solution found.")


def plot_results(periods, products, x, inv, demand):
    """Create visualization of production and inventory over time"""
    fig, (ax1, ax2, ax3) = plt.subplots(3, 1, figsize=(10, 12))

    # Production plot
    ind = np.arange(len(periods))
    width = 0.35

    prod_a = [x["A", t].value() for t in periods]
    prod_b = [x["B", t].value() for t in periods]

    ax1.bar(ind - width / 2, prod_a, width, label="Product A")
    ax1.bar(ind + width / 2, prod_b, width, label="Product B")
    ax1.set_ylabel("Units")
    ax1.set_title("Production Quantities")
    ax1.set_xticks(ind)
    ax1.set_xticklabels(periods)
    ax1.legend()

    # Inventory plot
    inv_a = [inv["A", t].value() for t in periods]
    inv_b = [inv["B", t].value() for t in periods]

    ax2.plot(periods, inv_a, "o-", label="Product A")
    ax2.plot(periods, inv_b, "s-", label="Product B")
    ax2.set_ylabel("Units")
    ax2.set_title("Ending Inventory Levels")
    ax2.legend()
    ax2.grid(True)

    # Production vs Demand
    demand_a = [demand["A", t] for t in periods]
    demand_b = [demand["B", t] for t in periods]

    ax3.plot(periods, prod_a, "o-", label="Production A")
    ax3.plot(periods, demand_a, "o--", label="Demand A")
    ax3.plot(periods, prod_b, "s-", label="Production B")
    ax3.plot(periods, demand_b, "s--", label="Demand B")
    ax3.set_ylabel("Units")
    ax3.set_title("Production vs Demand")
    ax3.legend()
    ax3.grid(True)

    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    main()

File: examples/multi-period.rs
#![allow(unused)]

use std::{collections::HashMap, time::Instant};
use good_lp::{
    constraint, default_solver, variable, variables, Expression, Solution, SolverModel
};

fn main() {
    // Start timing
    let start_time = Instant::now();
    
    // Initialize engine (as per template)
    engine::init();
    
    println!("Balance Engine - Multi-Period Production Planning Demo");
    
    // Define the problem data
    let products = vec!["A", "B"];
    let periods = vec!["January", "February", "March"];
    
    // Initial conditions
    let mut initial_inventory = HashMap::new();
    initial_inventory.insert("A", 100.0);
    initial_inventory.insert("B", 120.0);
    
    let mut safety_stock = HashMap::new();
    safety_stock.insert("A", 130.0);
    safety_stock.insert("B", 110.0);
    
    let mut production_cost = HashMap::new();
    production_cost.insert("A", 20.0);
    production_cost.insert("B", 25.0);
    
    let holding_cost_rate = 0.02;  // 2% of production cost
    
    // Calculate holding cost per unit per period
    let mut holding_cost = HashMap::new();
    for &p in &products {
        holding_cost.insert(p, production_cost[&p] * holding_cost_rate);
    }
    
    // Demand forecasts
    let mut demand = HashMap::new();
    demand.insert(("A", "January"), 700.0);
    demand.insert(("A", "February"), 900.0);
    demand.insert(("A", "March"), 1000.0);
    demand.insert(("B", "January"), 800.0);
    demand.insert(("B", "February"), 600.0);
    demand.insert(("B", "March"), 900.0);
    
    // Resource capacities
    let mut machine_capacity = HashMap::new();
    machine_capacity.insert("January", 3000.0);
    machine_capacity.insert("February", 2800.0);
    machine_capacity.insert("March", 3600.0);
    
    let mut labor_capacity = HashMap::new();
    labor_capacity.insert("January", 2500.0);
    labor_capacity.insert("February", 2300.0);
    labor_capacity.insert("March", 2400.0);
    
    // Resource requirements per unit
    let mut machine_hours = HashMap::new();
    machine_hours.insert("A", 1.5);
    machine_hours.insert("B", 1.6);
    
    let mut labor_hours = HashMap::new();
    labor_hours.insert("A", 1.1);
    labor_hours.insert("B", 1.2);
    
    // Create variables for the problem
    let mut vars = variables!();
    
    // x[p,t] = production of product p in period t
    let mut x = HashMap::new();
    for &p in &products {
        for &t in &periods {
            x.insert((p, t), vars.add(variable().min(0.0).integer().name(format!("x_{}_{}", p, t))));
        }
    }
    
    // inv[p,t] = inventory of product p at the end of period t
    let mut inv = HashMap::new();
    for &p in &products {
        for &t in &periods {
            inv.insert((p, t), vars.add(variable().min(0.0).integer().name(format!("inv_{}_{}", p, t))));
        }
    }
    
    // Objective function: minimize total cost (production + inventory holding)
    let mut objective = Expression::from(0.0);
    
    // Production costs
    for &p in &products {
        for &t in &periods {
            objective += production_cost[&p] * *x.get(&(p, t)).unwrap();
        }
    }
    
    // Holding costs
    for &p in &products {
        for &t in &periods {
            objective += holding_cost[&p] * *inv.get(&(p, t)).unwrap();
        }
    }
    
    // Create model with minimization objective
    let mut model = vars.minimise(objective.clone()).using(default_solver);
    
    // Add constraints
    
    // 1. Inventory balance constraints
    for &p in &products {
        for (i, &t) in periods.iter().enumerate() {
            if i == 0 {  // first period
                model = model.with(constraint!(
                    *inv.get(&(p, t)).unwrap() == initial_inventory[&p] + *x.get(&(p, t)).unwrap() - demand[&(p, t)]
                ));
            } else {
                let prev_t = periods[i - 1];
                model = model.with(constraint!(
                    *inv.get(&(p, t)).unwrap() == *inv.get(&(p, prev_t)).unwrap() + *x.get(&(p, t)).unwrap() - demand[&(p, t)]
                ));
            }
        }
    }
    
    // 2. Capacity constraints
    for &t in &periods {
        // Machine capacity
        let mut machine_usage = Expression::from(0.0);
        for &p in &products {
            machine_usage += machine_hours[&p] * *x.get(&(p, t)).unwrap();
        }
        model = model.with(constraint!(machine_usage <= machine_capacity[&t]));
        
        // Labor capacity
        let mut labor_usage = Expression::from(0.0);
        for &p in &products {
            labor_usage += labor_hours[&p] * *x.get(&(p, t)).unwrap();
        }
        model = model.with(constraint!(labor_usage <= labor_capacity[&t]));
    }
    
    // 3. Safety stock requirements (end of planning horizon)
    for &p in &products {
        let last_period = periods.last().unwrap();
        model = model.with(constraint!(*inv.get(&(p, last_period)).unwrap() >= safety_stock[&p]));
    }
    
    // Solve the model
    match model.solve() {
        Ok(solution) => {
            println!("\nOptimal Solution Found:");
            println!("Total Cost: €{:.2}", solution.eval(&objective));
            
            // Print production plan
            println!("\nProduction Plan:");
            println!("{:<10} {:<10} {:<10}", "Period", "Product A", "Product B");
            println!("{}", "-".repeat(30));
            for &t in &periods {
                println!("{:<10} {:<10} {:<10}", 
                         t, 
                         solution.value(*x.get(&("A", t)).unwrap()) as i32,
                         solution.value(*x.get(&("B", t)).unwrap()) as i32);
            }
            
            // Print ending inventory
            println!("\nEnding Inventory:");
            println!("{:<10} {:<10} {:<10}", "Period", "Product A", "Product B");
            println!("{}", "-".repeat(30));
            for &t in &periods {
                println!("{:<10} {:<10} {:<10}", 
                         t, 
                         solution.value(*inv.get(&("A", t)).unwrap()) as i32,
                         solution.value(*inv.get(&("B", t)).unwrap()) as i32);
            }
            
            // Calculate resource utilization
            println!("\nResource Utilization:");
            println!("{:<10} {:<30} {:<30}", 
                     "Period", 
                     "Machine Hours (Used/Available)", 
                     "Labor Hours (Used/Available)");
            println!("{}", "-".repeat(70));
            for &t in &periods {
                let mut machine_used = 0.0;
                let mut labor_used = 0.0;
                
                for &p in &products {
                    machine_used += machine_hours[&p] * solution.value(*x.get(&(p, t)).unwrap());
                    labor_used += labor_hours[&p] * solution.value(*x.get(&(p, t)).unwrap());
                }
                
                let machine_util = machine_used / machine_capacity[&t] * 100.0;
                let labor_util = labor_used / labor_capacity[&t] * 100.0;
                
                println!("{:<10} {:.1}/{} ({:.1}%) {:<5} {:.1}/{} ({:.1}%)",
                         t,
                         machine_used,
                         machine_capacity[&t],
                         machine_util,
                         "",
                         labor_used,
                         labor_capacity[&t],
                         labor_util);
            }
            
        },
        Err(e) => {
            println!("No optimal solution found. Error: {:?}", e);
        }
    }
    
    // Print elapsed time
    let elapsed = start_time.elapsed();
    println!("\nTime elapsed: {:?}", elapsed);
}